File: 31_Depth_Charge/python/depth_charge.py
"""
Original BASIC version as published in Basic Computer Games (1978)
https://www.atariarchives.org/basicgames/showpage.php?page=55

Converted to Python by Anson VanDoren in 2021
"""

import math
import random
from typing import Tuple


def show_welcome() -> None:
    # Clear screen. chr(27) is `Esc`, and the control sequence is
    # initiated by Ctrl+[
    # `J` is "Erase in Display" and `2J` means clear the entire screen
    print(chr(27) + "[2J")

    # Show the intro text, centered
    print("DEPTH CHARGE".center(45))
    print("Creative Computing  Morristown, New Jersey\n\n".center(45))


def get_num_charges() -> Tuple[int, int]:
    print("Depth Charge game\n")
    while True:
        search_area_str = input("Dimensions of search area? ")

        # Make sure the input is an integer
        try:
            search_area = int(search_area_str)
            break
        except ValueError:
            print("Must enter an integer number. Please try again...")

    num_charges = int(math.log2(search_area)) + 1
    return search_area, num_charges


def ask_for_new_game() -> None:
    answer = input("Another game (Y or N): ")
    if answer.lower().strip()[0] == "y":
        main()
    else:
        print("OK. Hope you enjoyed yourself")
        exit()


def show_shot_result(shot, location) -> None:
    result = "Sonar reports shot was "
    if shot[1] > location[1]:  # y-direction
        result += "north"
    elif shot[1] < location[1]:  # y-direction
        result += "south"
    if shot[0] > location[0]:  # x-direction
        result += "east"
    elif shot[0] < location[0]:  # x-direction
        result += "west"
    if shot[1] != location[1] or shot[0] != location[0]:
        result += " and "

    if shot[2] > location[2]:
        result += "too low."
    elif shot[2] < location[2]:
        result += "too high."
    else:
        result += "depth OK."
    print(result)
    return


def get_shot_input() -> Tuple[int, int, int]:
    while True:
        raw_guess = input("Enter coordinates: ")
        try:
            xyz = raw_guess.split()
        except ValueError:
            print("Please enter coordinates separated by spaces")
            print("Example: 3 2 1")
            continue
        try:
            x, y, z = (int(num) for num in xyz)
            return x, y, z
        except ValueError:
            print("Please enter whole numbers only")


def play_game(search_area, num_charges) -> None:
    print("\nYou are the captain of the destroyer USS Computer.")
    print("An enemy sub has been causing you trouble. Your")
    print(f"mission is to destroy it. You have {num_charges} shots.")
    print("Specify depth charge explosion point with a")
    print("trio of numbers -- the first two are the")
    print("surface coordinates; the third is the depth.")
    print("\nGood luck!\n")

    # Generate position for submarine
    a, b, c = (random.randint(0, search_area) for _ in range(3))

    # Get inputs until win or lose
    for i in range(num_charges):
        print(f"\nTrial #{i+1}")
        x, y, z = get_shot_input()

        if (x, y, z) == (a, b, c):
            print(f"\nB O O M ! ! You found it in {i+1} tries!\n")
            ask_for_new_game()
        else:
            show_shot_result((x, y, z), (a, b, c))

    # out of shots
    print("\nYou have been torpedoed! Abandon ship!")
    print(f"The submarine was at {a} {b} {c}")
    ask_for_new_game()


def main() -> None:
    search_area, num_charges = get_num_charges()
    play_game(search_area, num_charges)


if __name__ == "__main__":
    main()

File: 31_Depth_Charge/ruby/depthcharge.rb
#!/usr/bin/ruby

class DepthCharge
  def run_game
    output_title

    loop do
      puts "----------"
      print_instructions
      setup_game
      puts
      game_loop
      break unless get_input_another_game
    end

    puts "OK.  HOPE YOU ENJOYED YOURSELF."
  end

  def output_title
    puts "--- DEPTH CHARGE ---"
    puts "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY"
    puts
  end

  def get_input_y_or_n(message)
    loop do
      print message

      value = gets.chomp

      if value.downcase == "y"
        return true
      elsif value.downcase == "n"
        return false
      end

      puts "PLEASE ENTER Y/y OR N/n..."
      puts
    end
  end

  def get_input_positive_integer(message)
    loop do
      print message
      value = gets.chomp

      if value == "d"
        debug_game
        next
      end

      the_input = Integer(value) rescue 0

      if the_input < 1
        puts "PLEASE ENTER A POSITIVE NUMBER"
        puts
        next
      end

      return the_input
    end
  end

  def print_instructions
    puts <<~INSTRUCTIONS
      YOU ARE THE CAPTAIN OF THE DESTROYER USS COMPUTER
      AN ENEMY SUB HAS BEEN CAUSING YOU TROUBLE.  YOUR
      MISSION IS TO DESTROY IT.

      SPECIFY DEPTH CHARGE EXPLOSION POINT WITH A
      TRIO OF NUMBERS -- THE FIRST TWO ARE THE
      SURFACE COORDINATES (X, Y):
          WEST < X < EAST
          SOUTH < Y < NORTH

      THE THIRD IS THE DEPTH (Z):
        SHALLOW < Z < DEEP

      GOOD LUCK !

    INSTRUCTIONS
  end

  def debug_game
    puts "@enemy_x: %d" % @enemy_x
    puts "@enemy_y: %d" % @enemy_y
    puts "@enemy_z: %d" % @enemy_z
    puts "@num_tries: %d" % @num_tries
    puts "@trial: %d" % @trial
    puts
  end

  def setup_game
    @search_area_dimension = get_input_positive_integer("DIMENSION OF SEARCH AREA: ")

    @num_tries = Integer(Math.log(@search_area_dimension) / Math.log(2) + 1)
    setup_enemy
  end

  def setup_enemy
    @enemy_x = rand(1..@search_area_dimension)
    @enemy_y = rand(1..@search_area_dimension)
    @enemy_z = rand(1..@search_area_dimension)
  end

  def game_loop
    for @trial in 1..@num_tries do
      output_game_status()

      @shot_x = get_input_positive_integer("X: ")
      @shot_y = get_input_positive_integer("Y: ")
      @shot_z = get_input_positive_integer("Z: ")


      distance = (@enemy_x - @shot_x).abs +
        (@enemy_y - @shot_y).abs +
        (@enemy_z - @shot_z).abs

      if distance == 0
        you_win
        return
      else
        missed_shot
      end
    end

    puts

    you_lose
  end

  def output_game_status
    puts "YOU HAVE %d SHOTS REMAINING." % @num_tries - @trial + 1
    puts "TRIAL \#%d" % @trial
  end

  def you_win
    puts "\nB O O M ! ! YOU FOUND IT IN %d TRIES!" % @trial
    puts
  end

  def missed_shot
    missed_directions = []

    if @shot_x > @enemy_x
      missed_directions.push('TOO FAR EAST')
    elsif @shot_x < @enemy_x
      missed_directions.push('TOO FAR WEST')
    end

    if @shot_y > @enemy_y
      missed_directions.push('TOO FAR NORTH')
    elsif @shot_y < @enemy_y
      missed_directions.push('TOO FAR SOUTH')
    end

    if @shot_z > @enemy_z
      missed_directions.push('TOO DEEP')
    elsif @shot_z < @enemy_z
      missed_directions.push('TOO SHALLOW')
    end

    puts "SONAR REPORTS SHOT WAS: "
    puts "\t#{missed_directions.join("\n\t")}"
  end

  def you_lose
    puts "YOU HAVE BEEN TORPEDOED!  ABANDON SHIP!"
    puts "THE SUBMARINE WAS AT %d %d %d" % [@enemy_x, @enemy_y, @enemy_z]
  end

  def get_input_another_game
    return get_input_y_or_n("ANOTHER GAME (Y OR N): ")
  end
end

game = DepthCharge.new
game.run_game
